Reshape method (#1760)
LLNL/RAJA: docs/sphinx/user_guide/feature/view.rst

access array entries with stride N :subscript:`n` * N :subscript:`(n-1)` * ... * N :subscript:`(j+1)`.

Reshape
^^^^^^^

> The ``RAJA::Reshape`` method will provide developers a ``RAJA::View`` given a pointer, dimentions and a meta-layout.

Reviewer: ```suggestion
The ``RAJA::Reshape`` method takes a pointer to data and extents in each dimension as arguments and returns a ``RAJA::View`` object. The reshape method is specialized on a `meta-layout` template parameter, which specifies the stride ordering for the dimensions.

```

> Meta-layouts are predefined types which specify stride ordering. The first meta-layout is

Reviewer: ```suggestion
The first meta-layout is

```
after applying the change above.

based on a right-most layout which follows standard C indexing (right most index has unit stride).
The example below illustrates the following capability.

.. literalinclude:: ../../../../examples/reshape.cpp
   :start-after: _Reshape_right_start
   :end-before: _Reshape_right_end
   :language: C++
In the example above a ``RAJA::View`` is returned with right most indexing having unit stride.

The second supported meta-layout is the ``RAJA::layout_left`` which will follow a Fortran style indexing
in which the left most index has unit stride and the right most has the longest stride. The configuration
is demonstrated in the following example.

.. literalinclude:: ../../../../examples/reshape.cpp
   :start-after: _Reshape_left_start
   :end-before: _Reshape_left_end
   :language: C++

> Finally, users may specify a custom index ordering by providing a ``std::index_sequence`` list as a

Reviewer: ```suggestion
Finally, users may specify a custom index ordering by providing a ``std::index_sequence`` type as a

```

template argument.

.. literalinclude:: ../../../../examples/reshape.cpp
   :start-after: _Reshape_custom_start
   :end-before: _Reshape_custom_end
   :language: C++

> The example above will give unit stride to the right most index of the sequence while the left most index will have

Reviewer: The example code for this has entry '0' for the second entry in the sequence. I think that requires a sentence of explanation.

access array entries with stride N :subscript:`n` * N :subscript:`(n-1)` * ... * N :subscript:`(j+1)`.

Reshape
^^^^^^^
The ``RAJA::Reshape`` method is designed as a simplified multi-dimensional pointer abstraction

> only requiring the pointer, dimensions, and prefered layout type based on a template specilization.

Reviewer 2: This seems too focused on how its implemented via specialization than how its used. Could this just say that you need to provide a pointer, dimensions, and a layout type?

Author: Sure, I think we can make that works. I coined the term meta-layout as predefined layout types as suggested.

The first meta-layout is based on a right-most layout which follows standard C indexing (right most index has unit stride).
For example:

.. literalinclude:: ../../../../examples/reshape.cpp
   :start-after: _Reshape_right_start
   :end-before: _Reshape_right_end
   :language: C++
In the example above a ``RAJA::View`` is returned with right most indexing having unit stride.

The second supported meta-layout is the ``RAJA::layout_left`` which will follow a Fortran style indexing
in which the left most index has unit stride and the right most has the longest stride. The configuration
is demonstrated in the following example.

.. literalinclude:: ../../../../examples/reshape.cpp
   :start-after: _Reshape_left_start
   :end-before: _Reshape_left_end
   :language: C++

Finally, users may specify a custom index ordering by providing a ``std::index_sequence`` type as a
template argument for the ``RAJA::Reshape`` method. A ``RAJA::View`` enumerates the parathesis accessor
from left to right starting at zero. The ``std::index_sequence`` will hold an ordering of slowest (left most)
to fastest (right most) index for the ``RAJA::View``. In the following example, unit stride will be given to

> index 1 while index 2 is assumed to have the unit stride.

Reviewer 2: What is this saying? What's the parathesis accessor? Which has the unit stride?

Reviewer: I was confused by this also. The middle entry in the `std::index_sequence` is zero in the example, which requires explanation, IMO.

The first meta-layout is based on a right-most layout which follows standard C indexing (right most index has unit stride).
For example:

.. literalinclude:: ../../../../examples/reshape.cpp
   :start-after: _Reshape_right_start
   :end-before: _Reshape_right_end
   :language: C++
In the example above a ``RAJA::View`` is returned with right most indexing having unit stride.

The second supported meta-layout is the ``RAJA::layout_left`` which will follow a Fortran style indexing
in which the left most index has unit stride and the right most has the longest stride. The configuration
is demonstrated in the following example.

.. literalinclude:: ../../../../examples/reshape.cpp
   :start-after: _Reshape_left_start
   :end-before: _Reshape_left_end
   :language: C++

Finally, users may specify a custom index ordering by providing a ``std::index_sequence`` type as a
template argument to the ``RAJA::Reshape`` method. The members of the ``std::index_sequence`` enumerate
the striding order of the arguments of the ``RAJA::View`` parenthesis operator.

For example to use the C index ordering (``RAJA::layout_right``) with a 3-dimensional ``RAJA::View`` use the following 

> sequence type ``std::index_sequence<0U,1U,2U>``.  In the following example, we consider a permutation

Reviewer: I think this would be clearer if you added a sentence that states exactly what the indexing order is for this case.

^^^^^^^
The ``RAJA::Reshape`` method takes a pointer to data and extents in each dimension as arguments and returns a ``RAJA::View`` object. The reshape method is specialized on a `meta-layout` template parameter, which specifies the stride ordering for the dimensions.

The first meta-layout is

> 

Reviewer: Remove the blank line 100. Otherwise, Sphinx thinks line 99 is a separate paragraph.
